pkgs/top-level/all-packages-wrapper.nix

```diff
@@ -0,0 +1,270 @@
+      # decide, such as in case of functions, then we execute the
+      # "default" function with both arguments.
+      zipQuickFixAsPatches = path: pkgs: quickfix: whatif:
+        zipAttrsWith (name: values:
+          # Somebody added / removed a packaged in quickfix?
+          let pkgsName = concatStringsSep "." (path ++ [name]); in
+          # assert builtins.trace "zipQuickFixAsPatches (name: ${pkgsName})" true;
+          assert builtins.length values == 3;
+          let p = head values; q = head (tail values); w = head (tail (tail values)); in
+          if name == "pkgs" then q # We should not recurse in the top-level pkgs argument.
+          else if isAttrs p then
+            assert isAttrs q && isAttrs w; # Do not mutate the derivation
+            if isDerivation p then
+              assert isDerivation q && isDerivation w;
+              addErrorContext "While evaluating package ${pkgsName}" (quickFixAsPatches pkgsName p q w)
+            else
+              zipQuickFixAsPatches (path ++ [name]) p q w
+          else
+            q
+        ) [pkgs quickfix whatif];
+
+      # Additional list of packages which have ABI compatible fixes for the
+      # stable packages. These are used by abiCompatiblePatches.
+      quickfixPackages = {
+        allPackages = import ../../quickfix/pkgs/top-level/all-packages.nix;
+        aliasedPackages = import ../../quickfix/pkgs/top-level/all-packages-aliases.nix;
+      };
+
+      # Pipeline of modification involved to apply security patches:
+      #
+      #  1. We apply security patches on top of the current set of packages.
+      #
+      #  2. We check what package would be recompiled, if we were to
+      #     recompile instead of applying patches.
+      #
+      #  3. We only keep the set of packages where we only applied patches.
+      #
+      quickFix = pkgsWithOverrideFun quickfixPackages pkgs;
+      whatIf = pkgsWithOverrideFun quickfixPackages abiSec;
+      abiSec = zipQuickFixAsPatches ["pkgs"] pkgs quickFix whatIf;
+    in
+      abiSec;
+
+
+  # The package compositions.  Yes, this isn't properly indented.
+  pkgsFun = { allPackages, aliasedPackages }: pkgs:
+    let
+      defaultScope = pkgs // pkgs.xorg;
+      helperFunctions = pkgs_.stdenvAdapters // pkgs_.trivial-builders;
+      pkgsRet = helperFunctions // pkgs_;
+      pkgs_ = allPackages {
+        self = pkgs_;
+        inherit pkgs system crossSystem platform bootStdenv noSysDirs
+          gccWithCC gccWithProfiling config defaultScope helperFunctions;
+      };
+
+    in aliasFun { inherit allPackages aliasedPackages; } pkgsRet;
+
+  aliasFun = { allPackages, aliasedPackages }: pkgs:
+    let
+      aliases = aliasedPackages aliases pkgs;
+      tweakAlias = _n: alias: with lib;
+        if alias.recurseForDerivations or false then
+          removeAttrs alias ["recurseForDerivations"]
+        else alias;
+
+    in pkgs // lib.mapAttrs tweakAlias aliases;
```

Comment:
Ok, sorry for the comment spam. It finally occurred to me that the fixpoint is useful for aliases defind in terms of aliases (duh...).

Still don't get the security implication, but as I still need to find time to watch your talk, I'll defer judgement.